web/src/message_viewport.ts

```diff
@@ -1,30 +1,54 @@
 import $ from "jquery";
+import assert from "minimalistic-assert";
 
 import * as blueslip from "./blueslip";
 import * as message_lists from "./message_lists";
 import * as message_scroll_state from "./message_scroll_state";
+import type {Message} from "./message_store";
 import * as rows from "./rows";
 import * as util from "./util";
+import type {CachedValue} from "./util";
+
+type MessageViewportInfo = {
+    visible_top: number;
+    visible_bottom: number;
+    visible_height: number;
+};
+
+type DimenFunc = JQuery["height"] | JQuery["width"];
+type DimenWrapper = {
+    (): number;
+    (arg: string | number): JQuery;
+};
```

Comment:
Why isn't this type just `DimenFunc`?

Reply:
Ah I did this because original JQuery getter methods have the types of `number | undefined` but our wrapper dimension functions are safe and will always return a `number`. Therefore I defined our own safe type to avoid spamming couple of non-null assertions in other methods.

Reply:
OK, that's reasonable.